### PR TITLE
Allow to cancel the txdb upgrade via splashscreen keypress 'q'

### PR DESCRIPTION
**Pull request by jonasschnelli**
Progress calc credit: @sipa

This PR will accomplish three things:
* Show the progress of the UTXO-db upgrade in the Qt splashscreen and in the debug log:
  * `2017-06-23 07:49:55 Upgrading utxo-set database...`
  * `2017-06-23 07:49:55 [0%]...[0%]...[13%]...[25%]...[39%]...[52%]...[63%]...[76%]...[90%]...[DONE].`
* Allow to shutdown during the UTXO-db upgrade (ctrl-c, Qt quit)
* Allow to press `q` in splash screen to shutdown.

Let me explain why keypress instead of buttons:
The current splash screen is a drawn pixmap. Adding buttons there will require a complete re-design of the splashscreen (should be done once). It should be an auto-resizing QWidget as used in the other layout situations.

For getting a UTXO-upgrade abort into 0.15, this solution may be acceptable.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Speak up if you have a better, state-less solution for the UI progress callback.

**Comment by laanwj** [comment]
Awesome! Will test.

File: src/txdb.cpp [hunk 1]
```diff
@@ -319,13 +321,24 @@ bool CCoinsViewDB::Upgrade() {
...
     while (pcursor->Valid()) {
         boost::this_thread::interruption_point();
+        if (ShutdownRequested()) {
+            break;
+        }
         std::pair<unsigned char, uint256> key;
         if (pcursor->GetKey(key) && key.first == DB_COINS) {
+            if (count++ % 256 == 0) {
+                uint32_t high = 0x100 * *key.second.begin() + *(key.second.begin() + 1);
+                uiInterface.ShowProgress(_("Upgrading UTXO database") +"\n"+ _("(press q to shutdown and continue later"), (int)(high * 100.0 / 65536.0 + 0.5));
```
Comment (achow101) [inline src/txdb.cpp]: Couldn't this just be one string with a `\n` in the middle instead of two separate strings?

**Comment by sipa** [comment]
Tried this. Two issues:
* `2017-06-23 21:15:54 [15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...[15%]...`. We should probably only print at most once for each %.
* There seems to be a space missing between the message and the percentage in the spash screen.

File: src/txdb.cpp (hunk 1 above)
Comment (laanwj) [inline src/txdb.cpp]: From my experience, translators hate messages with html or newlines in them, increasing the chance to make a mistake there.

Comment (laanwj) [inline src/txdb.cpp]: `shut down` as two words maybe?

**Comment by jonasschnelli** [comment]
Rebased and overhauled.
Addresses @sipas nit.
Also reports now `[CANCELLED]` instead of the `[DONE]` when shutting down during txdb upgrade.

**Comment by laanwj** [comment]
Tested upgrade w/ my test data up to height=430234: 
- quitting with `q` worked, it indeed continues on next start
- killing with `kill -9` also made it continue at next start without problems
- progress logging in debug.log is as expected
- upgrade succeeded, node is running

File: src/qt/splashscreen.cpp [hunk 2]
```diff
@@ -131,13 +131,24 @@ SplashScreen::SplashScreen(Qt::WindowFlags f, const NetworkStyle *networkStyle)
...
 
 SplashScreen::~SplashScreen()
 {
     unsubscribeFromCoreSignals();
 }
 
+bool SplashScreen::eventFilter(QObject * obj, QEvent * ev) {
+    if (ev->type() == QEvent::KeyPress) {
+        QKeyEvent *keyEvent = static_cast<QKeyEvent *>(ev);
+        if(keyEvent->text()[0] == 'q' && continueAction != nullptr) {
```
Comment (laanwj) [inline src/qt/splashscreen.cpp]: I'd prefer `breakAction` to 'continueAction'.
Continue, to me, implies that it's part of a pause/continue kind of functionality, not early quit.

**Comment by jonasschnelli** [comment]
Agree with @laanwj, changed `continueAction` (and similar function names) to `breakAction`.
(amend commits / force push)

**Comment by laanwj** [comment]
ACK 542ce6e

**Comment by TheBlueMatt** [comment]
utACK with the changes at #10770.

**Pull request merged**